cmd/gke-certificates-controller/app/gke_approver.go

```diff
@@ -170,34 +189,138 @@ func hasExactUsages(csr *capi.CertificateSigningRequest, usages []capi.KeyUsage)
 		return false
 	}
 	return true
 }
 
-func isSelfNodeClientCert(csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
-	if !isNodeClientCert(csr, x509cr) {
+func isNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !isNodeCert(opts, csr, x509cr) {
+		return false
+	}
+	if !hasExactUsages(csr, kubeletServerUsages) {
 		return false
 	}
 	if csr.Spec.Username != x509cr.Subject.CommonName {
 		return false
 	}
 	return true
 }
+
+func validateNodeServerCert(opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	client := oauth2.NewClient(context.Background(), opts.tokenSource)
+
+	if err := validateNodeServerCertInner(client, opts, csr, x509cr); err != nil {
+		glog.Errorf("validating CSR %q: %v", csr.Name, err)
+		return false
+	}
+	return true
+}
+
+// Only check that DNS name and IP in SAN match an existing VM in the project.
+// Username was already checked against CN, so this CSR is coming from
+// authenticated kubelet.
+func validateNodeServerCertInner(client *http.Client, opts approverOptions, csr *capi.CertificateSigningRequest, x509cr *x509.CertificateRequest) error {
+	if len(x509cr.DNSNames) == 0 {
+		return errors.New("no SAN DNS names")
+	}
+	if len(x509cr.IPAddresses) == 0 {
+		return errors.New("no SAN IP addresses")
+	}
+	cs, err := compute.New(client)
+	if err != nil {
+		return fmt.Errorf("creating GCE API client: %v", err)
+	}
+
+outer:
+	for _, n := range x509cr.DNSNames {
+		inst, err := compute.NewInstancesService(cs).Get(opts.projectID, opts.zone, n).Do()
+		if err != nil {
+			return fmt.Errorf("DNS name %q doesn't match any VM in cluster project/zone: %v", n, err)
+		}
+		for _, iface := range inst.NetworkInterfaces {
+			for _, ip := range x509cr.IPAddresses {
+				if ip.String() == iface.NetworkIP {
+					continue outer
+				}
+			}
+		}
+		return fmt.Errorf("IP addresses in CSR (%q) don't match NetworkInterfaces on instance %q (%v)", x509cr.IPAddresses, n, inst.NetworkInterfaces)
+	}
+	return nil
+}
+
+type approverOptions struct {
+	projectID   string
+	zone        string
+	tokenSource oauth2.TokenSource
+}
+
+func loadApproverOptions(s *GKECertificatesController) (approverOptions, error) {
+	var a approverOptions
+
+	// Load gce.conf.
+	gceConfig := struct {
+		Global struct {
+			ProjectID string `gcfg:"project-id"`
+			TokenURL  string `gcfg:"token-url"`
+			TokenBody string `gcfg:"token-body"`
+		}
+	}{}
+	if err := gcfg.ReadFileInto(&gceConfig, s.GCEConfigPath); err != nil {
+		return a, err
+	}
+	a.projectID = gceConfig.Global.ProjectID
+	// Get the token source for GCE API.
+	a.tokenSource = gce.NewAltTokenSource(gceConfig.Global.TokenURL, gceConfig.Global.TokenBody)
+
+	// Get cluster zone from metadata server.
+	zone, err := metadata.Get("instance/attributes/cluster-location")
+	if err != nil {
+		return a, err
+	}
+	a.zone = strings.TrimSpace(zone)
```

Comment:
This isn't always going to be a zone. If it's a regional cluster, this will be a region.